src/common/dateInput/DateInputLogic.tsx

```diff
@@ -0,0 +1,240 @@
+  // keyboard
+  // 2) Blocks iOS devices from inputting characters other than numbers
+  pattern: '\\d*',
+};
+
+type DateObject = {
+  date?: number;
+  month?: number;
+  year?: number;
+};
+
+export interface InputComponentProps {
+  id: string;
+  name: string;
+  onChange: (event: React.FormEvent<HTMLInputElement>) => void;
+  value?: string;
+  innerRef?: React.RefObject<HTMLInputElement>;
+  min: number;
+  pattern: string;
+  type: string;
+  isInvalid: boolean;
+  label: string;
+}
+
+export interface WrapperComponentProps {
+  children?: React.ReactNode;
+  isInvalid: boolean;
+}
+
+interface Props {
+  onChange: (date: Date) => void;
+  wrapper: React.FC<WrapperComponentProps>;
+  value: Date | null;
+  isInvalid: boolean;
+  dateInputId: string;
+  monthInputId: string;
+  yearInputId: string;
+  dateInputName: string;
+  monthInputName: string;
+  yearInputName: string;
+  dateComponent: React.FC<InputComponentProps>;
+  monthComponent: React.FC<InputComponentProps>;
+  yearComponent: React.FC<InputComponentProps>;
+  dateInputLabel: string;
+  monthInputLabel: string;
+  yearInputLabel: string;
+}
+
+function getDateComponents(date: Date | null): DateObject {
+  if (!date) {
+    return {};
+  }
+
+  const monthDate = date.getDate();
+  const month = date.getMonth();
+  const year = date.getFullYear();
+
+  return {
+    date: monthDate,
+    month,
+    year,
+  };
+}
+
+function makeDate(date: DateObject): Date | null {
+  if (date.year && date.month && date.date) {
+    return new Date(date.year, date.month, date.date, 0, 0, 0, 0);
+  }
+
+  return null;
+}
+
+function getCharacterCount(number: number): number {
+  return number.toString().length;
+}
+
+function sanitizeInput(content: unknown) {
+  return String(content)
+    .split('')
+    .filter(character => DATE_INPUT_ALLOWED_CHARACTERS.includes(character))
+    .join('');
+}
+
+function cutToLength(string: string, length: number) {
+  if (string.length <= length) {
+    return string;
+  }
+
+  return string.slice(0, length);
+}
+
+function parseNumberFromValue(value: string, maxCharacterCount = 2): number {
+  // Removes all other characters than numbers from the result. This is
+  // a last fallback which makes negative numbers positive and removes
+  // possible characters from the output. The behavior of
+  // input[type="number"] is not consistent between platforms.
+  const sanitized = sanitizeInput(value);
+  const maxSized = cutToLength(sanitized, maxCharacterCount);
+
+  return Number(maxSized);
+}
+
+function DateInputLogic({
+  onChange,
+  value,
+  isInvalid,
+  wrapper,
+  dateInputId,
+  monthInputId,
+  yearInputId,
+  dateInputName,
+  monthInputName,
+  yearInputName,
+  dateComponent,
+  monthComponent,
+  yearComponent,
+  dateInputLabel,
+  monthInputLabel,
+  yearInputLabel,
+}: Props) {
+  const currentDate = getDateComponents(value);
+  const [cachedDate, setCachedDate] = React.useState<DateObject>(currentDate);
```

Comment:
is the date cached in the browser or is it only used in the state? If only state, I think another name could be more clear, but I can't think of an improvement so cached is fine